src/auth.rs

```diff
@@ -87,3 +90,166 @@ impl AppAuth {
+}
+
+/// Authenticate with Github's device flow. This starts the process to obtain a new `OAuth`.
+///
+/// See https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#device-flow for details.
+pub async fn authenticate_as_device<I, S>(client_id: SecretString, scope: I) -> Result<DeviceCodes>
+where
+    I: IntoIterator<Item = S>,
+    S: AsRef<str>,
+{
+    let crab = crate::OctocrabBuilder::new()
+        .base_url("https://github.com")?
+        .add_header(ACCEPT, "application/json".to_string())
+        .build()?;
+    let scope = {
+        let mut scopes = scope.into_iter();
+        let first = scopes
+            .next()
+            .map(|s| s.as_ref().to_string())
+            .unwrap_or_default();
+        scopes.fold(first, |i: String, n| i + "," + n.as_ref())
+    };
+    let mut r: DeviceCodes = crab
+        .post(
+            "login/device/code",
+            Some(&DeviceFlow {
+                client_id: client_id.expose_secret(),
+                scope: &scope,
+            }),
+        )
+        .await?;
+    r.crab = crab;
+    r.client_id = Some(client_id);
+    Ok(r)
+}
+
+/// The device codes as returned from step 1 of Github's device flow.
+///
+/// See https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#response-parameters
+#[derive(Deserialize, Clone)]
+pub struct DeviceCodes {
+    /// The device verification code is 40 characters and used to verify the device.
+    pub device_code: String,
+    /// The user verification code is displayed on the device so the user can enter the
+    /// code in a browser. This code is 8 characters with a hyphen in the middle.
+    pub user_code: String,
+    /// The verification URL where users need to enter the user_code: https://github.com/login/device.
+    pub verification_uri: String,
+    /// The number of seconds before the device_code and user_code expire. The default is
+    /// 900 seconds or 15 minutes.
+    pub expires_in: u64,
+    /// The minimum number of seconds that must pass before you can make a new access
+    /// token request (POST https://github.com/login/oauth/access_token) to complete the
+    /// device authorization. For example, if the interval is 5, then you cannot make a
+    /// new request until 5 seconds pass. If you make more than one request over 5
+    /// seconds, then you will hit the rate limit and receive a slow_down error.
+    pub interval: u64,
+
+    // Used to poll for the device codes.
+    #[serde(skip)]
+    crab: crate::Octocrab,
+
+    // Used for polling the device. This should be filled in as soon as possible.
+    #[serde(skip)]
+    client_id: Option<SecretString>,
+}
+
+impl DeviceCodes {
+    /// Poll Github to see if authentication codes are available.
+    ///
+    /// See `https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#response-parameters` for details.
+    pub async fn poll_once(&self) -> Result<Result<OAuth, Continue>> {
```

Comment:
This should be implemented on `Octocrab` and have device codes and client_id as arguments. This should also use `either::Either` for the inner type rather than result.

Reply:
I didn't want to add any dependencies on my PR, but I agree that this is better. I'll make that change.